File: packages/go/rpc.go
package erpc

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"sync/atomic"
	"time"
)

// BatchCall is one call in a caller-defined, unsplit JSON-RPC batch.
type BatchCall struct {
	Method string
	Params any
}

type batchPolicy uint8

const (
	batchAny batchPolicy = iota
	batchSolanaStandard
	batchUnsupported
)

// RPCNamespace provides generic and batch JSON-RPC access.
type RPCNamespace struct {
	transport *httpRPCTransport
	policy    batchPolicy
}

// Endpoint returns the public endpoint without credentials, query, or fragment.
func (n *RPCNamespace) Endpoint() string { return n.transport.endpoint.String() }

// Request executes one JSON-RPC call and decodes its result into result.
func (n *RPCNamespace) Request(ctx context.Context, method string, params any, result any) error {
	return n.transport.request(ctx, method, params, result)
}

// Raw is an alias for Request for forward-compatible methods.
func (n *RPCNamespace) Raw(ctx context.Context, method string, params any, result any) error {
	return n.Request(ctx, method, params, result)
}

// Batch sends exactly one JSON-RPC batch and returns results in caller order.
// The transport never silently splits a batch.
func (n *RPCNamespace) Batch(ctx context.Context, calls []BatchCall) ([]json.RawMessage, error) {
	if n.policy == batchUnsupported && len(calls) != 0 {
		return nil, sdkError(ErrorBatchPolicy, "leader RPC methods do not support batching")
	}
	if n.policy == batchSolanaStandard {
		if err := validateSolanaBatch(calls); err != nil {
			return nil, err
		}
	}
	return n.transport.batch(ctx, calls)
}

// PendingRequest is an inert typed request. I/O starts when Send is called.
type PendingRequest[T any] struct {
	namespace *RPCNamespace
	method    string
	params    any
}

// NewRequest creates a typed pending JSON-RPC request.
func NewRequest[T any](namespace *RPCNamespace, method string, params any) *PendingRequest[T] {
	return &PendingRequest[T]{namespace: namespace, method: method, params: params}
}

// Send performs the pending request.
func (r *PendingRequest[T]) Send(ctx context.Context) (T, error) {
	var result T
	err := r.namespace.Request(ctx, r.method, r.params, &result)
	return result, err
}

type httpRPCTransport struct {
	apiKey   string
	endpoint *url.URL
	headers  http.Header
	timeout  time.Duration
	client   *http.Client
	nextID   atomic.Uint64
	maxBatch int
}

type wireRequest struct {
	JSONRPC string `json:"jsonrpc"`
	ID      uint64 `json:"id"`
	Method  string `json:"method"`
	Params  any    `json:"params,omitempty"`
}

type wireResponse struct {
	JSONRPC string          `json:"jsonrpc"`
	ID      uint64          `json:"id"`
	Result  json.RawMessage `json:"result"`
	Error   *rpcErrorObject `json:"error"`
}

type rpcErrorObject struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

func newHTTPRPCTransport(config resolvedConfig, endpoint *url.URL) *httpRPCTransport {
	t := &httpRPCTransport{
		apiKey: config.apiKey, endpoint: endpoint, headers: config.headers.Clone(),
		timeout: config.timeout, client: config.httpClient, maxBatch: 256,
	}
	t.nextID.Store(0)
	return t
}

func (t *httpRPCTransport) id() uint64 { return t.nextID.Add(1) }

func (t *httpRPCTransport) request(ctx context.Context, method string, params, result any) error {
	id := t.id()
	var response wireResponse
	if err := t.post(ctx, wireRequest{JSONRPC: "2.0", ID: id, Method: method, Params: params}, &response); err != nil {
		return err
	}
	if response.ID != id {
		return sdkError(ErrorInvalidResponse, "ERPC returned an unexpected response id")
	}
	if response.Error != nil {
		return t.rpcError(response.Error)
	}
	if response.Result == nil {
		return sdkError(ErrorInvalidResponse, "ERPC returned an invalid response")
	}
	if err := json.Unmarshal(response.Result, result); err != nil {
		return sdkError(ErrorInvalidResponse, "ERPC returned an unexpected result shape")
	}
	return nil
}

func (t *httpRPCTransport) batch(ctx context.Context, calls []BatchCall) ([]json.RawMessage, error) {
	if len(calls) == 0 {
		return []json.RawMessage{}, nil
	}
	if len(calls) > t.maxBatch {
		return nil, sdkError(ErrorInvalidResponse, fmt.Sprintf("a batch may contain at most %d calls", t.maxBatch))
	}
	requests := make([]wireRequest, len(calls))
	order := make([]uint64, len(calls))
	for i, call := range calls {
		id := t.id()
		order[i] = id
		requests[i] = wireRequest{JSONRPC: "2.0", ID: id, Method: call.Method, Params: call.Params}
	}
	var responses []wireResponse
	if err := t.post(ctx, requests, &responses); err != nil {
		return nil, err
	}
	byID := make(map[uint64]wireResponse, len(responses))
	for _, response := range responses {
		if _, exists := byID[response.ID]; exists {
			return nil, sdkError(ErrorInvalidResponse, "ERPC returned a duplicate batch id")
		}
		byID[response.ID] = response
	}
	results := make([]json.RawMessage, len(order))
	for i, id := range order {
		response, exists := byID[id]
		if !exists {
			return nil, sdkError(ErrorInvalidResponse, "ERPC omitted a batch response")
		}
		if response.Error != nil {
			return nil, t.rpcError(response.Error)
		}
		if response.Result == nil {
			return nil, sdkError(ErrorInvalidResponse, "ERPC returned an invalid batch item")
		}
		results[i] = response.Result
		delete(byID, id)
	}
	if len(byID) != 0 {
		return nil, sdkError(ErrorInvalidResponse, "ERPC returned an unexpected batch response id")
	}
	return results, nil
}

func (t *httpRPCTransport) post(ctx context.Context, body, destination any) error {
	payload, err := json.Marshal(body)
	if err != nil {
		return sdkError(ErrorConfig, "RPC parameters could not be serialized")
	}
	u := *t.endpoint
	query := u.Query()
	query.Set("api-key", t.apiKey)
	u.RawQuery = query.Encode()
	requestContext, cancel := context.WithTimeout(ctx, t.timeout)
	defer cancel()
	request, err := http.NewRequestWithContext(requestContext, http.MethodPost, u.String(), bytes.NewReader(payload))
	if err != nil {
		return sdkError(ErrorConfig, "unable to create RPC request")
	}
	request.Header = t.headers.Clone()
	request.Header.Set("Content-Type", "application/json")
	response, err := t.client.Do(request)
	if err != nil {
		if requestContext.Err() != nil {
			return requestContext.Err()
		}
		return sdkError(ErrorTransport, "unable to reach ERPC")
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return &Error{Kind: ErrorHTTP, Message: "request failed", Status: response.StatusCode}
	}
	decoder := json.NewDecoder(io.LimitReader(response.Body, 16<<20))
	if err := decoder.Decode(destination); err != nil {
		return sdkError(ErrorInvalidResponse, "ERPC returned malformed JSON")
	}
	return nil
}

func (t *httpRPCTransport) rpcError(value *rpcErrorObject) error {
	return &Error{
		Kind: ErrorRPC, Code: value.Code,
		Message: redactCredential(value.Message, t.apiKey),
		Data:    redactJSON(value.Data, t.apiKey),
	}
}

func validateSolanaBatch(calls []BatchCall) error {
	heavy := map[string]bool{
		"getPriorityFeeEstimate": true, "getProgramAccounts": true,
		"getProgramAccountsV2": true, "getTokenLargestAccounts": true,
	}
	var hasHeavy, hasStandard bool
	for _, call := range calls {
		if heavy[call.Method] {
			hasHeavy = true
		} else {
			hasStandard = true
		}
	}
	if hasHeavy && hasStandard {
		return sdkError(ErrorBatchPolicy, "Solana indexed and standard RPC methods cannot share a batch")
	}
	return nil
}

func decodeRaw[T any](raw json.RawMessage) (T, error) {
	var result T
	if err := json.Unmarshal(raw, &result); err != nil {
		return result, sdkError(ErrorInvalidResponse, "ERPC returned an unexpected result shape")
	}
	return result, nil
}

func boolString(value bool) string { return strconv.FormatBool(value) }

File: packages/rust/src/rpc.rs
use std::{
    collections::{HashMap, HashSet},
    marker::PhantomData,
    sync::{
        Arc,
        atomic::{AtomicU64, Ordering},
    },
    time::Duration,
};

use reqwest::{Client, header::HeaderMap};
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use serde_json::Value;
use tokio_util::sync::CancellationToken;
use url::Url;

use crate::{
    ErpcError, JsonRpcErrorObject, Result,
    error::{invalid_response, rpc_error, timeout},
};

/// JSON-RPC request or response identifier.
#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
#[serde(untagged)]
pub enum JsonRpcId {
    /// Numeric identifier used by the built-in transports.
    Number(u64),
    /// String identifier accepted from compatible servers.
    String(String),
}

/// One call in an unsplit JSON-RPC batch.
#[derive(Clone, Debug, Serialize)]
pub struct RpcBatchCall {
    /// Wire-compatible JSON-RPC method name.
    pub method: String,
    /// Positional array or named object parameters.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

impl RpcBatchCall {
    /// Creates a call and serializes its parameters.
    pub fn new(method: impl Into<String>, params: impl Serialize) -> Result<Self> {
        Ok(Self {
            method: method.into(),
            params: Some(serde_json::to_value(params).map_err(|_| {
                ErpcError::Config("RPC parameters could not be serialized".to_owned())
            })?),
        })
    }

    /// Creates a parameterless call.
    #[must_use]
    pub fn without_params(method: impl Into<String>) -> Self {
        Self {
            method: method.into(),
            params: None,
        }
    }
}

/// Per-request cancellation options.
#[derive(Clone, Debug, Default)]
pub struct RequestOptions {
    /// Cancellation token observed while connecting and awaiting a response.
    pub cancellation: Option<CancellationToken>,
}

/// Configuration for the HTTP JSON-RPC transport.
#[derive(Clone)]
pub struct HttpTransportConfig {
    /// ERPC API key.
    pub api_key: String,
    /// HTTP endpoint, without the API key query parameter.
    pub endpoint: Url,
    /// Extra request headers.
    pub headers: HeaderMap,
    /// Maximum accepted caller batch size.
    pub max_batch_size: usize,
    /// Request timeout.
    pub timeout: Duration,
    /// Configured HTTP client.
    pub client: Client,
}

/// Credential-safe HTTP JSON-RPC transport.
pub struct HttpJsonRpcTransport {
    api_key: String,
    endpoint: Url,
    headers: HeaderMap,
    max_batch_size: usize,
    timeout: Duration,
    client: Client,
    next_id: AtomicU64,
}

impl HttpJsonRpcTransport {
    /// Creates a transport. Prefer [`crate::ErpcClient`] for normal use.
    #[must_use]
    pub fn new(config: HttpTransportConfig) -> Self {
        let mut endpoint = config.endpoint;
        endpoint.set_query(None);
        endpoint.set_fragment(None);
        Self {
            api_key: config.api_key,
            endpoint,
            headers: config.headers,
            max_batch_size: config.max_batch_size,
            timeout: config.timeout,
            client: config.client,
            next_id: AtomicU64::new(1),
        }
    }

    /// Public endpoint without credentials, query, or fragment.
    #[must_use]
    pub fn endpoint(&self) -> String {
        self.endpoint.to_string()
    }

    /// Maximum number of calls in one server batch.
    #[must_use]
    pub const fn max_batch_size(&self) -> usize {
        self.max_batch_size
    }

    /// Sends one JSON-RPC request.
    pub async fn request<T: DeserializeOwned>(
        &self,
        method: &str,
        params: Option<Value>,
        options: &RequestOptions,
    ) -> Result<T> {
        let id = self.id();
        let request = WireRequest {
            jsonrpc: "2.0",
            id: id.clone(),
            method,
            params,
        };
        let response = self.post(&request, options).await?;
        self.unwrap(&response, &id)
    }

    /// Sends one intact JSON-RPC batch and returns results in caller order.
    pub async fn batch(
        &self,
        calls: &[RpcBatchCall],
        options: &RequestOptions,
    ) -> Result<Vec<Value>> {
        if calls.is_empty() {
            return Ok(Vec::new());
        }
        if calls.len() > self.max_batch_size {
            return Err(invalid_response(format!(
                "A batch may contain at most {} calls",
                self.max_batch_size
            )));
        }
        let requests: Vec<_> = calls
            .iter()
            .map(|call| WireRequest {
                jsonrpc: "2.0",
                id: self.id(),
                method: call.method.as_str(),
                params: call.params.clone(),
            })
            .collect();
        let response = self.post(&requests, options).await?;
        let Value::Array(responses) = response else {
            if let Ok(failure) = serde_json::from_value::<WireFailure>(response) {
                return Err(rpc_error(failure.error, &self.api_key));
            }
            return Err(invalid_response("ERPC returned a non-array batch response"));
        };
        let mut by_id = HashMap::new();
        for response in responses {
            let id = response
                .get("id")
                .cloned()
                .ok_or_else(|| invalid_response("ERPC returned an invalid batch item"))?;
            let id: JsonRpcId = serde_json::from_value(id)
                .map_err(|_| invalid_response("ERPC returned an invalid batch id"))?;
            if by_id.insert(id, response).is_some() {
                return Err(invalid_response("ERPC returned a duplicate batch id"));
            }
        }
        requests
            .iter()
            .map(|request| {
                let response = by_id.remove(&request.id).ok_or_else(|| {
                    invalid_response(format!("ERPC omitted batch response id {}", request.id))
                })?;
                self.unwrap(&response, &request.id)
            })
            .collect()
    }

    fn id(&self) -> JsonRpcId {
        let id = self.next_id.fetch_add(1, Ordering::Relaxed);
        if id == u64::MAX {
            self.next_id.store(1, Ordering::Relaxed);
        }
        JsonRpcId::Number(id)
    }

    async fn post(&self, body: &impl Serialize, options: &RequestOptions) -> Result<Value> {
        let mut url = self.endpoint.clone();
        url.query_pairs_mut().append_pair("api-key", &self.api_key);
        let request = self
            .client
            .post(url)
            .headers(self.headers.clone())
            .json(body)
            .send();
        let result = if let Some(cancellation) = &options.cancellation {
            tokio::select! {
                () = cancellation.cancelled() => return Err(ErpcError::Aborted),
                result = tokio::time::timeout(self.timeout, request) => result,
            }
        } else {
            tokio::time::timeout(self.timeout, request).await
        };
        let response = result
            .map_err(|_| timeout(self.timeout))?
            .map_err(|_| ErpcError::Transport("Unable to reach ERPC".to_owned()))?;
        if !response.status().is_success() {
            return Err(ErpcError::Http {
                status: response.status().as_u16(),
            });
        }
        let text = response
            .text()
            .await
            .map_err(|_| ErpcError::Transport("Unable to read ERPC response".to_owned()))?;
        serde_json::from_str(&text).map_err(|_| invalid_response("ERPC returned malformed JSON"))
    }

    fn unwrap<T: DeserializeOwned>(&self, response: &Value, expected: &JsonRpcId) -> Result<T> {
        let object = response
            .as_object()
            .ok_or_else(|| invalid_response("ERPC returned an invalid response"))?;
        let id: JsonRpcId = serde_json::from_value(
            object
                .get("id")
                .cloned()
                .ok_or_else(|| invalid_response("ERPC returned an invalid response"))?,
        )
        .map_err(|_| invalid_response("ERPC returned an invalid response id"))?;
        if &id != expected {
            return Err(invalid_response("ERPC returned an unexpected response id"));
        }
        if let Some(error) = object.get("error") {
            let error: JsonRpcErrorObject = serde_json::from_value(error.clone())
                .map_err(|_| invalid_response("ERPC returned an invalid RPC error"))?;
            return Err(rpc_error(error, &self.api_key));
        }
        let result = object
            .get("result")
            .cloned()
            .ok_or_else(|| invalid_response("ERPC returned an invalid response"))?;
        serde_json::from_value(result)
            .map_err(|_| invalid_response("ERPC returned an unexpected result shape"))
    }
}

impl std::fmt::Display for JsonRpcId {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Number(id) => id.fmt(formatter),
            Self::String(id) => id.fmt(formatter),
        }
    }
}

#[derive(Serialize)]
struct WireRequest<'a> {
    jsonrpc: &'static str,
    id: JsonRpcId,
    method: &'a str,
    #[serde(skip_serializing_if = "Option::is_none")]
    params: Option<Value>,
}

#[derive(Deserialize)]
struct WireFailure {
    error: JsonRpcErrorObject,
}

/// An inert typed request. Network I/O begins only when `send` is awaited.
pub struct PendingRpcRequest<T> {
    transport: Arc<HttpJsonRpcTransport>,
    method: String,
    params: Option<Value>,
    marker: PhantomData<fn() -> T>,
}

impl<T: DeserializeOwned> PendingRpcRequest<T> {
    /// Sends with default options.
    pub async fn send(self) -> Result<T> {
        self.send_with(RequestOptions::default()).await
    }

    /// Sends with cancellation options.
    pub async fn send_with(self, options: RequestOptions) -> Result<T> {
        self.transport
            .request(&self.method, self.params, &options)
            .await
    }
}

/// An inert batch request that preserves one caller batch on the wire.
pub struct PendingRpcBatchRequest {
    transport: Arc<HttpJsonRpcTransport>,
    calls: Vec<RpcBatchCall>,
}

impl PendingRpcBatchRequest {
    /// Sends with default options.
    pub async fn send(self) -> Result<Vec<Value>> {
        self.send_with(RequestOptions::default()).await
    }

    /// Sends with cancellation options.
    pub async fn send_with(self, options: RequestOptions) -> Result<Vec<Value>> {
        self.transport.batch(&self.calls, &options).await
    }
}

#[derive(Clone, Copy)]
pub(crate) enum BatchPolicy {
    Any,
    SolanaStandard,
    Unsupported,
}

/// Typed request builder over one JSON-RPC endpoint and parameter mode.
#[derive(Clone)]
pub struct RpcNamespace {
    transport: Arc<HttpJsonRpcTransport>,
    policy: BatchPolicy,
}

impl RpcNamespace {
    pub(crate) const fn new(transport: Arc<HttpJsonRpcTransport>, policy: BatchPolicy) -> Self {
        Self { transport, policy }
    }

    /// Public endpoint without credentials.
    #[must_use]
    pub fn endpoint(&self) -> String {
        self.transport.endpoint()
    }

    /// Creates a typed request using wire-compatible method and parameters.
    pub fn request<T: DeserializeOwned>(
        &self,
        method: impl Into<String>,
        params: impl Serialize,
    ) -> Result<PendingRpcRequest<T>> {
        Ok(PendingRpcRequest {
            transport: Arc::clone(&self.transport),
            method: method.into(),
            params: Some(serde_json::to_value(params).map_err(|_| {
                ErpcError::Config("RPC parameters could not be serialized".to_owned())
            })?),
            marker: PhantomData,
        })
    }

    /// Creates a typed parameterless request.
    #[must_use]
    pub fn request_without_params<T: DeserializeOwned>(
        &self,
        method: impl Into<String>,
    ) -> PendingRpcRequest<T> {
        PendingRpcRequest {
            transport: Arc::clone(&self.transport),
            method: method.into(),
            params: None,
            marker: PhantomData,
        }
    }

    /// Alias for arbitrary methods not yet represented by a convenience method.
    pub fn raw<T: DeserializeOwned>(
        &self,
        method: impl Into<String>,
        params: Option<Value>,
    ) -> PendingRpcRequest<T> {
        PendingRpcRequest {
            transport: Arc::clone(&self.transport),
            method: method.into(),
            params,
            marker: PhantomData,
        }
    }

    /// Validates and creates one unsplit batch request.
    pub fn batch(&self, calls: Vec<RpcBatchCall>) -> Result<PendingRpcBatchRequest> {
        if matches!(self.policy, BatchPolicy::Unsupported) {
            if !calls.is_empty() {
                return Err(ErpcError::BatchPolicy(
                    "Leader RPC methods do not support batching".to_owned(),
                ));
            }
        } else if matches!(self.policy, BatchPolicy::SolanaStandard) {
            validate_solana_standard_batch(&calls)?;
        }
        Ok(PendingRpcBatchRequest {
            transport: Arc::clone(&self.transport),
            calls,
        })
    }
}

fn validate_solana_standard_batch(calls: &[RpcBatchCall]) -> Result<()> {
    let heavy: HashSet<&str> = [
        "getPriorityFeeEstimate",
        "getProgramAccounts",
        "getProgramAccountsV2",
        "getTokenLargestAccounts",
    ]
    .into_iter()
    .collect();
    let contains_heavy = calls
        .iter()
        .any(|call| heavy.contains(call.method.as_str()));
    let contains_standard = calls
        .iter()
        .any(|call| !heavy.contains(call.method.as_str()));
    if contains_heavy && contains_standard {
        return Err(ErpcError::BatchPolicy(
            "Solana indexed and standard RPC methods cannot share a batch".to_owned(),
        ));
    }
    Ok(())
}
